### ISSUE TITLE
compilation failed for package ‘httpgd’ 

### ISSUE THREAD

**Issue by f-jx** [nx10/httpgd#73]
linux version: 5.4.72-microsoft-standard-WSL2
conda version:4.10.3
R version:4.1.0

the error tips:

Selection: 25
try URL’https://mirrors.sustech.edu.cn/CRAN/src/contrib/httpgd_1.1.1.tar.gz'
Content type 'application/octet-stream' length 179789 bytes (175 KB)
==================================================
downloaded 175 KB

* installing *source* package ‘httpgd’ ...
** package ‘httpgd’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
x86_64-conda-linux-gnu-c++ -std=gnu++17 -I"/opt/miniconda3/lib/R/include" -DNDEBUG -Ilib -DBOOST_NO_AUTO_PTR -DFMT_HEADER_ONLY -I'/opt/miniconda3/lib/R/library/cpp11/include' -I'/opt/miniconda3/lib/R/library/BH/include' -I'/opt/miniconda3/lib/R/library/later/include' -I'/opt/miniconda3/lib/R/library/systemfonts/include' -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/miniconda3/include -I/opt/miniconda3/include -Wl,-rpath-link,/opt/miniconda3/lib   -fpic  -fvisibility-inlines-hidden  -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/miniconda3/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/r-base-split_1621590585644/work=/usr/local/src/conda/r-base-4.1.0 -fdebug-prefix-map=/opt/miniconda3=/usr/local/src/conda-prefix  -c AsyncLater.cpp -o AsyncLater.o
In file included from /opt/miniconda3/x86_64-conda-linux-gnu/include/c++/9.3.0/chrono:41,
                 from /opt/miniconda3/x86_64-conda-linux-gnu/include/c++/9.3.0/mutex:39,
                 from AsyncLater.cpp:4:
/opt/miniconda3/x86_64-conda-linux-gnu/include/c++/9.3.0/ctime:80:11: error: '::timespec_get' has not been declared
   80 |   using ::timespec_get;
      |           ^~~~~~~~~~~~
make: *** [/opt/miniconda3/lib/R/etc/Makeconf:177：AsyncLater.o] error
ERROR: compilation failed for package ‘httpgd’
* removing ‘/opt/miniconda3/lib/R/library/httpgd’

The downloaded source packages are in
        ‘/tmp/RtmpPu9rzA/downloaded_packages’
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Warning message:
In install.packages("httpgd") :
  installation of package ‘httpgd’ had non-zero exit status

**Comment by f-jx** [comment]
Thanks, I final installed the httpgd successfully
I reinstall R 4.0, and libpng-dev in my WSL, and it's OK.!

**Comment by QiangShiPKU** [comment]
> Thanks, I final installed the httpgd successfully
> I reinstall R 4.0, and libpng-dev in my WSL, and it's OK.!

the same compilation error for R 4.0.5 and R 4.1.1., but not including R 4.0.0

**Comment by nx10** [comment]
> > Thanks, I final installed the httpgd successfully
> > I reinstall R 4.0, and libpng-dev in my WSL, and it's OK.!
> 
> the same compilation error for R 4.0.5 and R 4.1.1., but not including R 4.0.0

Could you please post the error log and system information?

**Comment by QiangShiPKU** [comment]
> > > Thanks, I final installed the httpgd successfully
> > > I reinstall R 4.0, and libpng-dev in my WSL, and it's OK.!
> > 
> > 
> > the same compilation error for R 4.0.5 and R 4.1.1., but not including R 4.0.0
> 
> Could you please post the error log and system information?

OS: Ubuntu 18.04.2

* installing *source* package ‘httpgd’ ...
** package ‘httpgd’ successfully unpacked and MD5 sums checked
** using staged installation
Found pkg-config cflags and libs!
Using PKG_CFLAGS=-I/raid1/shiq/app/miniconda3/envs/R/include/cairo -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/glib-2.0 -I/raid1/shiq/app/miniconda3/envs/R/lib/glib-2.0/include -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/pixman-1 -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/uuid -I/raid1/shiq/app/miniconda3/envs/R/include/libxml2 -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/libpng16 -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/freetype2
Using PKG_LIBS=-L/raid1/shiq/app/miniconda3/envs/R/lib -lcairo -lfreetype
** libs
rm -f httpgd.so AsyncLater.o Base64.o DrawData.o Httpgd.o HttpgdApiAsync.o HttpgdDataStore.o HttpgdDev.o HttpgdWebServer.o PlotHistory.o RendererCairo.o RendererJson.o RendererManager.o RendererMeta.o RendererStrings.o RendererSvg.o RendererTikZ.o cpp11.o devGeneric.o
x86_64-conda-linux-gnu-c++ -std=gnu++17 -I"/raid1/shiq/app/miniconda3/envs/R/lib/R/include" -DNDEBUG -I/raid1/shiq/app/miniconda3/envs/R/include/cairo -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/glib-2.0 -I/raid1/shiq/app/miniconda3/envs/R/lib/glib-2.0/include -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/pixman-1 -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/uuid -I/raid1/shiq/app/miniconda3/envs/R/include/libxml2 -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/libpng16 -I/raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include/freetype2 -Ilib -DBOOST_NO_AUTO_PTR -DFMT_HEADER_ONLY -I'/raid1/shiq/app/miniconda3/envs/R/lib/R/library/cpp11/include' -I'/raid1/shiq/app/miniconda3/envs/R/lib/R/library/BH/include' -I'/raid1/shiq/app/miniconda3/envs/R/lib/R/library/later/include' -I'/raid1/shiq/app/miniconda3/envs/R/lib/R/library/systemfonts/include' -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /raid1/shiq/app/miniconda3/envs/R/include -I/raid1/shiq/app/miniconda3/envs/R/include -Wl,-rpath-link,/raid1/shiq/app/miniconda3/envs/R/lib   -fpic  -fvisibility-inlines-hidden  -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /raid1/shiq/app/miniconda3/envs/R/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/r-base-split_1630154298835/work=/usr/local/src/conda/r-base-4.1.1 -fdebug-prefix-map=/raid1/shiq/app/miniconda3/envs/R=/usr/local/src/conda-prefix  -c AsyncLater.cpp -o AsyncLater.o
In file included from /raid1/shiq/app/miniconda3/envs/R/x86_64-conda-linux-gnu/include/c++/9.3.0/chrono:41,
                 from /raid1/shiq/app/miniconda3/envs/R/x86_64-conda-linux-gnu/include/c++/9.3.0/mutex:39,
                 from AsyncLater.cpp:4:
/raid1/shiq/app/miniconda3/envs/R/x86_64-conda-linux-gnu/include/c++/9.3.0/ctime:80:11: error: '::timespec_get' has not been declared
   80 |   using ::timespec_get;
      |           ^~~~~~~~~~~~
make: *** [/raid1/shiq/app/miniconda3/envs/R/lib/R/etc/Makeconf:177: AsyncLater.o] Error 1
ERROR: compilation failed for package ‘httpgd’
* removing ‘/raid1/shiq/app/miniconda3/envs/R/lib/R/library/httpgd’
* restoring previous ‘/raid1/shiq/app/miniconda3/envs/R/lib/R/library/httpgd’

**Comment by nx10** [comment]
I don't think this is a httpgd error but there is something wrong with the system (in your case anaconda) environment.

I found [this](https://root-forum.cern.ch/t/error-timespec-get-has-not-been-declared-with-conda-root-package/45712/8) thread which might help you.

```
conda upgrade -c conda-forge --all
```

Let me know if this works for you.

**Comment by QiangShiPKU** [comment]
> I don't think this is a httpgd error but there is something wrong with the system (in your case anaconda) environment.
> 
> I found [this](https://root-forum.cern.ch/t/error-timespec-get-has-not-been-declared-with-conda-root-package/45712/8) thread which might help you.
> 
> ```
> conda upgrade -c conda-forge --all
> ```
> 
> Let me know if this works for you.

This does not work but reminds me. I have made it by the following steps:
1) In my conda environment, I found all packages are from conda-forge channel, except these four pairs
<img width="538" alt="截屏2022-01-19 下午9 19 21" src="https://user-images.githubusercontent.com/40571792/150138762-7fe408b6-0179-4c81-a71c-398519bd17bf.png">

2) I created a new conda environment and installed R 4.1.1 and httpgd. According to this new environment, I updated packages above to v9.4.0 and installed httpgd in my original environment successfully
<img width="544" alt="截屏2022-01-19 下午9 19 28" src="https://user-images.githubusercontent.com/40571792/150140446-47000e68-dbf4-4f7a-be33-eb5ef8d0f236.png">

Thank you very much for your help!
